File: Core/Architect.API.Core.DataAccess/_Generated/Attachments.vb
Imports Architect.Common.Proxy

Namespace Architect.API.Core.DataAccess.General

    ''' <summary>
    ''' 
    ''' </summary>
    Partial Public NotInheritable Class Attachments

        ''' <summary>
        ''' Crea un registro en la tabla Attachments.
        ''' </summary>
        ''' <param name="attachmentsItem">Instancia de Attachments</param>
        ''' <returns>Instancia de Attachments</returns>
        Public Shared Function Create(attachmentsItem As Architect.API.Core.Contracts.General.Attachments) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("INSERT INTO Attachments " &
                                              "(Id, DocumentType, Description, FileName, FileSize, FileContent, EntityType, EntityId, CompanyId, UpdateUserCode, UpdateDate) " &
                                        "VALUES(@:Id, @:DocumentType, @:Description, @:FileName, @:FileSize, @:FileContent, @:EntityType, @:EntityId, @:CompanyId, @:UpdateUserCode, @:UpdateDate)", "Attachments", "Research")
                .AddParameter("Id", DbType.Decimal, 9, attachmentsItem.Id.IsEmpty(), attachmentsItem.Id)
                .AddParameter("DocumentType", DbType.Decimal, 5, attachmentsItem.DocumentType.IsEmpty(), attachmentsItem.DocumentType)
                .AddParameter("Description", DbType.AnsiString, 120, attachmentsItem.Description.IsEmpty(), attachmentsItem.Description)
                .AddParameter("FileName", DbType.AnsiString, 80, attachmentsItem.FileName.IsEmpty(), attachmentsItem.FileName)
                .AddParameter("FileSize", DbType.Decimal, 9, attachmentsItem.FileSize.IsEmpty(), attachmentsItem.FileSize)
                If attachmentsItem.FileContent.IsNotEmpty Then
                    Dim contenido As Byte() = Common.Helpers.ByteHandler.FileToBytes(attachmentsItem.FileContent)
                    .AddParameter("FileContent", DbType.Binary, 0, False, contenido)
                Else
                    .AddParameter("FileContent", DbType.Binary, 0, True, Nothing)
                End If
                .AddParameter("EntityType", DbType.Decimal, 5, attachmentsItem.EntityType.IsEmpty(), attachmentsItem.EntityType)
                .AddParameter("EntityId", DbType.Decimal, 18, attachmentsItem.EntityId.IsEmpty(), attachmentsItem.EntityId)
                .AddParameter("CompanyId", DbType.Decimal, 5, attachmentsItem.CompanyId.IsEmpty(), attachmentsItem.CompanyId)
                .AddParameter("UpdateUserCode", DbType.Decimal, 9, attachmentsItem.UpdateUserCode.IsEmpty(), attachmentsItem.UpdateUserCode)
                .AddParameter("UpdateDate", DbType.DateTime, 0, attachmentsItem.UpdateDate.IsEmpty(), attachmentsItem.UpdateDate)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera un registro en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <param name="id"></param>
        ''' <returns>Instancia de Attachments</returns>
        Public Shared Function Retrieve(id As Integer) As Architect.API.Core.Contracts.General.Attachments
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.General.Attachments = Nothing
            With New DataManagerFactory("SELECT Id, DocumentType, Description, FileName, FileSize, FileContent, EntityType, EntityId, Attachments.CompanyId, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " &
                                          "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode " &
                                         "WHERE Id=@:Id", "Attachments", "Research")
                .AddParameter("Id", DbType.Decimal, 9, False, id)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapper(rows.Rows(0))
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Recupera una lista de registros en la tabla Attachments.
        ''' </summary>
        ''' <returns>Lista de instancias de Attachments</returns>
        Public Shared Function RetrieveAll() As List(Of Architect.API.Core.Contracts.General.Attachments)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Architect.API.Core.Contracts.General.Attachments)
            With New DataManagerFactory("SELECT Id, DocumentType, Description, FileName, FileSize, FileContent, EntityType, EntityId, Attachments.CompanyId, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " &
                                          "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode", "Attachments", "Research")
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(DataRowMapper(itemRow))
                    Next
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Verifica si existe un registro en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <returns>Verdadero si existe el registro, en caso contrario falso.</returns>
        Public Shared Function RetrieveLastKey() As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT NVL(MAX(Id),0) " &
                                          "FROM Attachments", "Attachments", "Research")
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Actualiza un registro en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <param name="attachmentsItem">Instancia de Attachments</param>
        ''' <returns>Cantidad de registros actualizados.</returns>
        Public Shared Function Update(attachmentsItem As Architect.API.Core.Contracts.General.Attachments) As Integer
            Dim rows As Integer = 0
            If attachmentsItem.FileContent.IsNotEmpty Then
                Dim contenido As Byte() = Common.Helpers.ByteHandler.FileToBytes(attachmentsItem.FileContent)
                With New DataManagerFactory("UPDATE Attachments " &
                                               "SET DocumentType=@:DocumentType, Description=@:Description, FileName=@:FileName, FileSize=@:FileSize, FileContent=@:FileContent, EntityType=@:EntityType, EntityId=@:EntityId, CompanyId=@:CompanyId, UpdateUserCode=@:UpdateUserCode, UpdateDate=@:UpdateDate " &
                                             "WHERE Id=@:Id", "Attachments", "Research")
                    .AddParameter("DocumentType", DbType.Decimal, 5, attachmentsItem.DocumentType.IsEmpty(), attachmentsItem.DocumentType)
                    .AddParameter("Description", DbType.AnsiString, 120, attachmentsItem.Description.IsEmpty(), attachmentsItem.Description)
                    .AddParameter("FileName", DbType.AnsiString, 80, attachmentsItem.FileName.IsEmpty(), attachmentsItem.FileName)
                    .AddParameter("FileSize", DbType.Decimal, 9, attachmentsItem.FileSize.IsEmpty(), attachmentsItem.FileSize)

                    .AddParameter("FileContent", DbType.Binary, 0, False, contenido)
                    .AddParameter("EntityType", DbType.Decimal, 5, attachmentsItem.EntityType.IsEmpty(), attachmentsItem.EntityType)
                    .AddParameter("EntityId", DbType.Decimal, 18, attachmentsItem.EntityId.IsEmpty(), attachmentsItem.EntityId)
                    .AddParameter("CompanyId", DbType.Decimal, 5, attachmentsItem.CompanyId.IsEmpty(), attachmentsItem.CompanyId)
                    .AddParameter("UpdateUserCode", DbType.Decimal, 9, attachmentsItem.UpdateUserCode.IsEmpty(), attachmentsItem.UpdateUserCode)
                    .AddParameter("UpdateDate", DbType.DateTime, 0, attachmentsItem.UpdateDate.IsEmpty(), attachmentsItem.UpdateDate)
                    .AddParameter("Id", DbType.Decimal, 9, False, attachmentsItem.Id)
                    rows = .CommandExecute()
                End With
            Else
                With New DataManagerFactory("UPDATE Attachments " &
                                               "SET DocumentType=@:DocumentType, Description=@:Description, FileName=@:FileName, FileSize=@:FileSize, EntityType=@:EntityType, EntityId=@:EntityId, CompanyId=@:CompanyId, UpdateUserCode=@:UpdateUserCode, UpdateDate=@:UpdateDate " &
                                             "WHERE Id=@:Id", "Attachments", "Research")
                    .AddParameter("DocumentType", DbType.Decimal, 5, attachmentsItem.DocumentType.IsEmpty(), attachmentsItem.DocumentType)
                    .AddParameter("Description", DbType.AnsiString, 120, attachmentsItem.Description.IsEmpty(), attachmentsItem.Description)
                    .AddParameter("FileName", DbType.AnsiString, 80, attachmentsItem.FileName.IsEmpty(), attachmentsItem.FileName)
                    .AddParameter("FileSize", DbType.Decimal, 9, attachmentsItem.FileSize.IsEmpty(), attachmentsItem.FileSize)
                    .AddParameter("EntityType", DbType.Decimal, 5, attachmentsItem.EntityType.IsEmpty(), attachmentsItem.EntityType)
                    .AddParameter("EntityId", DbType.Decimal, 18, attachmentsItem.EntityId.IsEmpty(), attachmentsItem.EntityId)
                    .AddParameter("CompanyId", DbType.Decimal, 5, attachmentsItem.CompanyId.IsEmpty(), attachmentsItem.CompanyId)
                    .AddParameter("UpdateUserCode", DbType.Decimal, 9, attachmentsItem.UpdateUserCode.IsEmpty(), attachmentsItem.UpdateUserCode)
                    .AddParameter("UpdateDate", DbType.DateTime, 0, attachmentsItem.UpdateDate.IsEmpty(), attachmentsItem.UpdateDate)
                    .AddParameter("Id", DbType.Decimal, 9, False, attachmentsItem.Id)
                    rows = .CommandExecute()
                End With
            End If

            Return rows
        End Function

        ''' <summary>
        ''' Elimina un registro en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <param name="id"></param>
        ''' <returns>Cantidad de registros eliminados.</returns>
        Public Shared Function Delete(id As Integer) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("DELETE FROM Attachments " &
                                         "WHERE Id=@:Id", "Attachments", "Research")
                .AddParameter("Id", DbType.Decimal, 9, False, id)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera la cantidad de registros existentes en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <param name="id"></param>
        ''' <returns>Cantidad de registros encontrados.</returns>
        Public Shared Function Count(id As Integer) As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT COUNT(Id) " &
                                          "FROM Attachments " &
                                         "WHERE Id=@:Id", "Attachments", "Research")
                .AddParameter("Id", DbType.Decimal, 9, False, id)
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera una lista de registros en la tabla Attachments.
        ''' </summary>
        ''' <param name="entityid"></param>
        ''' <param name="entitytype"></param>
        ''' <param name="companyid"></param>
        ''' <returns>Lista de instancias de Attachments</returns>
        Public Shared Function RetrieveByEntityIdEntityTypeCompanyId(entityid As Long, entitytype As Integer, companyid As Integer) As List(Of Architect.API.Core.Contracts.General.Attachments)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Architect.API.Core.Contracts.General.Attachments)
            With New DataManagerFactory("Select Id, DocumentType, Description, FileName, FileSize, FileContent, EntityType, EntityId, Attachments.CompanyId, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " &
                                          "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode " &
                                         "WHERE EntityId=@:EntityId AND EntityType=@:EntityType AND CompanyId=@:CompanyId", "Attachments", "Research")
                .AddParameter("EntityId", DbType.Decimal, 18, False, entityid)
                .AddParameter("EntityType", DbType.Decimal, 5, False, entitytype)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyid)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(DataRowMapper(itemRow))
                    Next
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Elimina un registro en la tabla Attachments por medio de su clave primaria.
        ''' </summary>
        ''' <param name="entityid"></param>
        ''' <param name="entitytype"></param>
        ''' <param name="companyid"></param>
        ''' <returns>Cantidad de registros eliminados.</returns>
        Public Shared Function DeleteByEntityIdEntityTypeCompanyId(entityid As Long, entitytype As Integer, companyid As Integer) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("DELETE FROM Attachments " &
                                         "WHERE EntityId=@:EntityId AND EntityType=@:EntityType AND CompanyId=@:CompanyId", "Attachments", "Research")
                .AddParameter("EntityId", DbType.Decimal, 18, False, entityid)
                .AddParameter("EntityType", DbType.Decimal, 5, False, entitytype)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyid)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

#Region "Helper"

        ''' <summary>
        ''' Permite hacer el mapeo entre una fila de un datatable hacia una instancia de Attachments.
        ''' </summary>
        ''' <returns>Instancia de Attachments</returns>
        Private Shared Function DataRowMapper(itemRow As DataRow) As Architect.API.Core.Contracts.General.Attachments
            Return New Architect.API.Core.Contracts.General.Attachments With {.Id = itemRow.NumericValue("Id"), .DocumentType = itemRow.NumericValue("DocumentType"), .Description = itemRow.StringValue("Description"), .FileName = itemRow.StringValue("FileName"), .FileSize = itemRow.NumericValue("FileSize"), .FileContent = itemRow.StringValue("FileContent"), .EntityType = itemRow.NumericValue("EntityType"), .EntityId = itemRow.NumericValue("EntityId"), .CompanyId = itemRow.NumericValue("CompanyId"), .UpdateUserCode = itemRow.NumericValue("UpdateUserCode"), .UpdateUserName = itemRow.StringValue("UpdateUserName"), .UpdateDate = itemRow.DateTimeValue("UpdateDate")}
        End Function

#End Region

    End Class

End Namespace

File: Core/Architect.API.Core.DataAccess/_Generated/Attachments.cs
using Architect.DataFactory;
using Architect.Utilities.Extensions;
using System;
using System.Collections.Generic;
using System.Data;
using DbType = Architect.DataFactory.Enumerations.DbType;

namespace Architect.API.Core.DataAccess.General
{
    /// <summary>
    /// Archivos adjuntos.
    /// </summary>
    public sealed partial class Attachments
    {

        /// <summary>
        /// Crea un registro en la tabla Attachments.
        /// </summary>
        /// <param name="attachmentsItem">Instancia de Attachments</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros creados.</returns>
        public static int Create(Architect.API.Core.Contracts.General.Attachments attachmentsItem, IDbConnection connection = null)
        {
            if (attachmentsItem.UpdateDate.IsEmpty())
            {
                attachmentsItem.UpdateDate = DateTime.Now;
            }
            return Database.Insert("INSERT INTO Attachments (Id, CompanyId, EntityType, EntitySubType, EntityId, DocumentType, Description, FileName, FileSize, FileContent, UpdateUserCode, UpdateDate) " +
                                                 "VALUES(:Id, :CompanyId, :EntityType, :EntitySubType, :EntityId, :DocumentType, :Description, :FileName, :FileSize, :FileContent, :UpdateUserCode, :UpdateDate)")
                            .AddParameter("Id", DbType.Decimal, 9, attachmentsItem.Id)
                            .AddParameter("CompanyId", DbType.Decimal, 5, attachmentsItem.CompanyId)
                            .AddParameter("EntityType", DbType.Decimal, 5, attachmentsItem.EntityType)
                            .AddParameter("EntitySubType", DbType.Decimal, 5, attachmentsItem.EntitySubType)
                            .AddParameter("EntityId", DbType.Decimal, 18, attachmentsItem.EntityId)
                            .AddParameter("DocumentType", DbType.Decimal, 5, attachmentsItem.DocumentType)
                            .AddParameter("Description", DbType.AnsiString, 120, attachmentsItem.Description)
                            .AddParameter("FileName", DbType.AnsiString, 80, attachmentsItem.FileName)
                            .AddParameter("FileSize", DbType.Decimal, 9, attachmentsItem.FileSize)
                            .AddParameter("FileContent", DbType.AnsiString, 9, attachmentsItem.FileContent)
                            .AddParameter("UpdateUserCode", DbType.Decimal, 9, attachmentsItem.UpdateUserCode)
                            .AddParameter("UpdateDate", DbType.DateTime, 0, attachmentsItem.UpdateDate)
                            .Execute(connection, "Research");
        }

        /// <summary>
        /// Crea una lista de registros en la tabla Attachments.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="attachmentsItems">Lista de instancia de Attachments</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Create(List<Architect.API.Core.Contracts.General.Attachments> attachmentsItems, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (Architect.API.Core.Contracts.General.Attachments item in attachmentsItems)
            {
                result.Add(Create(item, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Recupera un registro en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <param name="id">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Instancia de Attachments</returns>
        public static Architect.API.Core.Contracts.General.Attachments Retrieve(int id, int companyId, IDbConnection connection = null)
        {
            Architect.API.Core.Contracts.General.Attachments result = null;
            Database.Select("SELECT Id, Attachments.CompanyId, EntityType, EntitySubType, EntityId, DocumentType, Description, FileName, FileSize, FileContent, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " +
                              "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode " +
                             "WHERE Attachments.Id=:Id AND Attachments.CompanyId=:CompanyId")
                        .AddParameter("Id", DbType.Decimal, 9, id)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result = DataReaderToAttachments(reader);
                        }));
            return result;
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Attachments.
        /// </summary>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="filter">Filtro personalizado.</param>
        /// <param name="parameters">Lista de parámetros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista de instancias de Attachments</returns>
        public static List<Architect.API.Core.Contracts.General.Attachments> RetrieveAll(int companyId, string filter, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            List<Architect.API.Core.Contracts.General.Attachments> result = new List<Architect.API.Core.Contracts.General.Attachments>();
            Database.Select("SELECT Id, Attachments.CompanyId, EntityType, EntitySubType, EntityId, DocumentType, Description, FileName, FileSize, FileContent, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " +
                              "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode " +
                             "WHERE Attachments.CompanyId=:CompanyId" + filter)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .AddParameter(parameters)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result.Add(DataReaderToAttachments(reader));
                        }));
            return result;
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Attachments.
        /// </summary>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="filter">Filtro personalizado.</param>
        /// <param name="beginIndex">Indice inicial para el paginado.</param>
        /// <param name="endIndex">Indice final para el paginado.</param>
        /// <param name="parameters">Lista de parámetros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista de instancias de Attachments</returns>
        public static List<Architect.API.Core.Contracts.General.Attachments> RetrieveAll(int companyId, string filter, int beginIndex, int endIndex, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            List<Architect.API.Core.Contracts.General.Attachments> result = new List<Architect.API.Core.Contracts.General.Attachments>();
            if (beginIndex == 0)
            {
                beginIndex = 1;
            }
            if (endIndex == 0 || endIndex < beginIndex)
            {
                endIndex = int.MaxValue;
            }
            Database.Select("SELECT * FROM (" +
                            "SELECT Id, Attachments.CompanyId, EntityType, EntitySubType, EntityId, DocumentType, Description, FileName, FileSize, FileContent, Attachments.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Attachments.UpdateDate " +
                                   ", ROW_NUMBER() OVER (ORDER BY Attachments.Id DESC) RowNumber " +
                              "FROM Attachments LEFT JOIN UserMember um ON um.UserId = Attachments.UpdateUserCode " +
                             "WHERE Attachments.CompanyId=:CompanyId" + filter +
                               ") WHERE RowNumber BETWEEN :beginIndex AND :endIndex")
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .AddParameter(parameters)
                        .AddParameter("beginIndex", DbType.Decimal, 9, beginIndex)
                        .AddParameter("endIndex", DbType.Decimal, 9, endIndex)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result.Add(DataReaderToAttachments(reader));
                        }));
            return result;
        }

        /// <summary>
        /// Genera complemento de la consulta para habilitar el filtro por columnas establecidas.
        /// </summary>
        /// <param name="filter">Permite filtrar la lista de registros.</param>
        /// <param name="includeWhere">Indica que el complemento debe incluir el WHERE del comando.</param>
        /// <returns>Complemento de la consulta</returns>
        public static string FilterBuilder(string filter, bool includeWhere = true)
        {
            string result = string.Empty;


            return result;
        }

        /// <summary>
        /// Último valor asignado a clave única de la tabla Attachments.
        /// </summary>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Último valor asignado.</returns>
        public static int RetrieveLastKey(IDbConnection connection = null)
        {

            return (int)Database.Select("SELECT NVL(MAX(Id),0) " +
                                     "FROM Attachments")
                                .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Actualiza un registro en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <param name="attachmentsItem">Instancia de Attachments</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros actualizados.</returns>
        public static int Update(Architect.API.Core.Contracts.General.Attachments attachmentsItem, IDbConnection connection = null)
        {
            if (attachmentsItem.UpdateDate.IsEmpty())
            {
                attachmentsItem.UpdateDate = DateTime.Now;
            }
            return Database.Update("UPDATE Attachments " +
                                      "SET CompanyId=:CompanyId, EntityType=:EntityType, EntitySubType=:EntitySubType, EntityId=:EntityId, DocumentType=:DocumentType, Description=:Description, FileName=:FileName, FileSize=:FileSize, FileContent=:FileContent, UpdateUserCode=:UpdateUserCode, UpdateDate=:UpdateDate " +
                                    "WHERE Id=:Id")
                                .AddParameter("CompanyId", DbType.Decimal, 5, attachmentsItem.CompanyId)
                                .AddParameter("EntityType", DbType.Decimal, 5, attachmentsItem.EntityType)
                                .AddParameter("EntitySubType", DbType.Decimal, 5, attachmentsItem.EntitySubType)
                                .AddParameter("EntityId", DbType.Decimal, 18, attachmentsItem.EntityId)
                                .AddParameter("DocumentType", DbType.Decimal, 5, attachmentsItem.DocumentType)
                                .AddParameter("Description", DbType.AnsiString, 120, attachmentsItem.Description)
                                .AddParameter("FileName", DbType.AnsiString, 80, attachmentsItem.FileName)
                                .AddParameter("FileSize", DbType.Decimal, 9, attachmentsItem.FileSize)
                                .AddParameter("FileContent", DbType.AnsiString, 9, attachmentsItem.FileContent)
                                .AddParameter("UpdateUserCode", DbType.Decimal, 9, attachmentsItem.UpdateUserCode)
                                .AddParameter("UpdateDate", DbType.DateTime, 0, attachmentsItem.UpdateDate)
                                .AddParameter("Id", DbType.Decimal, 9, attachmentsItem.Id)
                                .Execute(connection, "Research");
        }

        /// <summary>
        /// Actualiza una lista de registros en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="attachmentsItems">Lista de instancia de Attachments</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Update(List<Architect.API.Core.Contracts.General.Attachments> attachmentsItems, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (Architect.API.Core.Contracts.General.Attachments item in attachmentsItems)
            {
                result.Add(Update(item, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Elimina un registro en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <param name="id">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros eliminados.</returns>
        public static int Delete(int id, int companyId, IDbConnection connection = null)
        {
            return Database.Delete("DELETE FROM Attachments " +
                                    "WHERE Id=:Id AND CompanyId=:CompanyId")
                                .AddParameter("Id", DbType.Decimal, 9, id)
                                .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                                .Execute(connection, "Research");
        }

        /// <summary>
        /// Actualiza una lista de registros en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="idList">Lista de Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Delete(List<int> idList, int companyId, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (int item in idList)
            {
                result.Add(Delete(item, companyId, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Recupera la cantidad de registros existentes en la tabla Attachments por medio de su clave primaria.
        /// </summary>
        /// <param name="id">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros encontrados.</returns>
        public static int Count(int id, int companyId, IDbConnection connection = null)
        {

            return (int)Database.Select("SELECT COUNT(Id) " +
                                          "FROM Attachments " +
                                         "WHERE Id=:Id AND CompanyId=:CompanyId")
                               .AddParameter("Id", DbType.Decimal, 9, id)
                               .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                               .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Recupera la cantidad de registros existentes en la tabla Attachments que cumplen con el filtro.
        /// </summary>
        /// <param name="filter">Permite filtrar la lista de registros</param>
        /// <param name="parameters">Lista de parámetros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros encontrados.</returns>
        public static int Count(string filter, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            return (int)Database.Select("SELECT COUNT(Id) " +
                                          "FROM Attachments " + filter)
                               .AddParameter(parameters)
                               .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Convierte una fila de un 'DataReader' a un clase de tipo 'Attachments'.
        /// </summary>
        /// <param name="reader">DataReader</param>
        /// <param name="item">Instancia pre creada</param>
        /// <returns>Instancia con la información de la fila del 'DataReader'</returns>
        public static Architect.API.Core.Contracts.General.Attachments DataReaderToAttachments(System.Data.IDataReader reader, Architect.API.Core.Contracts.General.Attachments item = null)
        {
            if (item == null)
            {
                item = new Architect.API.Core.Contracts.General.Attachments();
            }
            item.Id = reader.IntegerValue("Id");
            item.CompanyId = reader.IntegerValue("CompanyId");
            item.EntityType = reader.IntegerValue("EntityType");
            item.EntitySubType = reader.IntegerValue("EntitySubType");
            item.EntityId = reader.Integer64Value("EntityId");
            item.DocumentType = reader.IntegerValue("DocumentType");
            item.Description = reader.StringValue("Description");
            item.FileName = reader.StringValue("FileName");
            item.FileSize = reader.IntegerValue("FileSize");
            item.FileContent = reader.StringValue("FileContent");
            item.UpdateUserCode = reader.IntegerValue("UpdateUserCode");
            item.UpdateUserName = reader.StringValue("UpdateUserName");
            item.UpdateDate = reader.DateTimeValue("UpdateDate");
            return item;
        }

    }

}

